core/src/main/java/org/elasticsearch/gateway/GatewayModule.java

```diff
@@ -19,19 +19,53 @@
 
 package org.elasticsearch.gateway;
 
-import org.elasticsearch.common.inject.*;
+import org.elasticsearch.common.inject.AbstractModule;
+import org.elasticsearch.common.settings.Settings;
+
+import java.util.HashMap;
+import java.util.Map;
 
 /**
  *
  */
 public class GatewayModule extends AbstractModule {
 
+    public static final String GATEWAY_TYPE_KEY = "gateway.type";
+
+    private final Map<String, Class<? extends Gateway>> gatewayTypes = new HashMap<>();
```

Comment:
If this is just for tests (why would someone want to have a custom gateway?), then I think we should use a pkg-private override here. See #12863 for examples how this was done for other mocks.